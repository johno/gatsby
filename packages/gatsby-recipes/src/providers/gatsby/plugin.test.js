const fs = require(`fs-extra`)
const path = require(`path`)

const plugin = require(`./plugin`)
const { addPluginToConfig, getPluginsFromConfig } = require(`./plugin`)
const resourceTestHelper = require(`../resource-test-helper`)

const root = path.join(__dirname, `./fixtures/gatsby-starter-blog`)
const helloWorldRoot = path.join(
  __dirname,
  `./fixtures/gatsby-starter-hello-world`
)
const name = `gatsby-plugin-foo`
const configPath = path.join(root, `gatsby-config.js`)

describe(`gatsby-plugin resource`, () => {
  test(`e2e plugin resource test`, async () => {
    await resourceTestHelper({
      resourceModule: plugin,
      resourceName: `GatsbyPlugin`,
      context: { root },
      initialObject: { id: name, name },
      partialUpdate: { id: name },
    })
  })

  test(`e2e plugin resource test with hello world starter`, async () => {
    await resourceTestHelper({
      resourceModule: plugin,
      resourceName: `GatsbyPlugin`,
      context: { root: helloWorldRoot },
      initialObject: { id: name, name },
      partialUpdate: { id: name },
    })
  })

  test(`does not add the same plugin twice by default`, async () => {
    const configSrc = await fs.readFile(configPath, `utf8`)
    const newConfigSrc = addPluginToConfig(
      configSrc,
      `gatsby-plugin-react-helmet`
    )
    const plugins = getPluginsFromConfig(newConfigSrc)

    const result = [...new Set(plugins)]

    expect(result).toEqual(plugins)
  })

  // A key isn't required for gatsby plugin, but when you want to distinguish
  // between multiple of the same plugin, you can specify it to target config changes.
  test(`validates the gatsby-source-filesystem specifies a key`, async () => {
    const result = plugin.validate({ name: `gatsby-source-filesystem` })

    expect(result.error).toEqual(
      `gatsby-source-filesystem requires a key to be set`
    )
  })

  test(`adds multiple gatsby-source-filesystems and reads with key`, async () => {
    const context = { root: helloWorldRoot }
    const fooPlugin = {
      key: `foo-data-sourcing`,
      name: `gatsby-source-filesystem`,
      options: {
        name: `foo`,
        path: `foo`,
      },
    }
    const barPlugin = {
      key: `bar-data-sourcing`,
      name: `gatsby-source-filesystem`,
      options: {
        name: `bar`,
        path: `bar`,
      },
    }

    await plugin.create(context, fooPlugin)
    await plugin.create(context, barPlugin)

    const barResult = await plugin.read(context, barPlugin.key)
    const fooResult = await plugin.read(context, fooPlugin.key)

    expect(barResult.key).toEqual(barPlugin.key)
    expect(fooResult.key).toEqual(fooPlugin.key)
    expect(barResult.options.name).toEqual(barPlugin.options.name)
    expect(fooResult.options.name).toEqual(fooPlugin.options.name)

    const newBarResult = await plugin.update(context, {
      ...barResult,
      options: { path: `new-bar` },
    })

    expect(newBarResult.key).toEqual(barPlugin.key)
    expect(newBarResult.options).toEqual({ path: `new-bar` })

    await plugin.destroy(context, barResult)
    await plugin.destroy(context, fooResult)
  })

  test(`handles config options as an object`, async () => {
    const configSrc = await fs.readFile(configPath, `utf8`)
    const newConfigSrc = addPluginToConfig(configSrc, `gatsby-plugin-foo`, {
      foo: 1,
      bar: `baz`,
      baz: `qux`,
      otherStuff: [
        {
          foo: `bar2`,
          bar: [{ foo: `bar` }],
        },
      ],
    })

    const result = getPluginsFromConfig(newConfigSrc)

    expect(result).toMatchSnapshot()
  })
})
