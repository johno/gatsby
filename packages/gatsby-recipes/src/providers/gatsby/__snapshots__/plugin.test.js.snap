// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gatsby-plugin resource e2e plugin resource test with hello world starter: GatsbyPlugin create 1`] = `
Object {
  "_message": "Installed gatsby-plugin-foo in gatsby-config.js",
  "id": "gatsby-plugin-foo",
  "name": "gatsby-plugin-foo",
  "options": undefined,
}
`;

exports[`gatsby-plugin resource e2e plugin resource test with hello world starter: GatsbyPlugin create plan 1`] = `
Object {
  "currentState": "/**
 * Configure your Gatsby site with this file.
 *
 * See: https://www.gatsbyjs.org/docs/gatsby-config/
 */
module.exports = {
  /* Your site config here */
  plugins: [],
}
",
  "describe": "Install gatsby-plugin-foo in gatsby-config.js",
  "diff": "- Original  - 1
+ Modified  + 1

@@ -5,6 +5,6 @@
   */
  module.exports = {
    /* Your site config here */
-   plugins: [],
+   plugins: [\\"gatsby-plugin-foo\\"],
  }
",
  "id": "gatsby-plugin-foo",
  "name": "gatsby-plugin-foo",
  "newState": "/**
 * Configure your Gatsby site with this file.
 *
 * See: https://www.gatsbyjs.org/docs/gatsby-config/
 */
module.exports = {
  /* Your site config here */
  plugins: [\\"gatsby-plugin-foo\\"],
}
",
}
`;

exports[`gatsby-plugin resource e2e plugin resource test with hello world starter: GatsbyPlugin destroy 1`] = `undefined`;

exports[`gatsby-plugin resource e2e plugin resource test with hello world starter: GatsbyPlugin update 1`] = `
Object {
  "_message": "Installed gatsby-plugin-foo in gatsby-config.js",
  "id": "gatsby-plugin-foo",
  "name": "gatsby-plugin-foo",
  "options": undefined,
}
`;

exports[`gatsby-plugin resource e2e plugin resource test with hello world starter: GatsbyPlugin update plan 1`] = `
Object {
  "currentState": "/**
 * Configure your Gatsby site with this file.
 *
 * See: https://www.gatsbyjs.org/docs/gatsby-config/
 */
module.exports = {
  /* Your site config here */
  plugins: [\\"gatsby-plugin-foo\\"],
}
",
  "describe": "Install gatsby-plugin-foo in gatsby-config.js",
  "diff": "Compared values have no visual difference.",
  "id": "gatsby-plugin-foo",
  "name": "gatsby-plugin-foo",
  "newState": "/**
 * Configure your Gatsby site with this file.
 *
 * See: https://www.gatsbyjs.org/docs/gatsby-config/
 */
module.exports = {
  /* Your site config here */
  plugins: [\\"gatsby-plugin-foo\\"],
}
",
}
`;

exports[`gatsby-plugin resource e2e plugin resource test: GatsbyPlugin create 1`] = `
Object {
  "_message": "Installed gatsby-plugin-foo in gatsby-config.js",
  "id": "gatsby-plugin-foo",
  "name": "gatsby-plugin-foo",
  "options": undefined,
}
`;

exports[`gatsby-plugin resource e2e plugin resource test: GatsbyPlugin create plan 1`] = `
Object {
  "currentState": "const redish = \`#c5484d\`
module.exports = {
  siteMetadata: {
    title: \`Bricolage\`,
    author: \`Kyle Mathews\`,
    homeCity: \`San Francisco\`,
  },
  plugins: [
    {
      resolve: \`gatsby-source-filesystem\`,
      options: {
        path: \`\${__dirname}/src/pages\`,
        name: \`pages\`,
      },
    },
    \`gatsby-transformer-sharp\`,
    \`gatsby-plugin-emotion\`,
    {
      resolve: \`gatsby-plugin-typography\`,
      options: {
        pathToConfigModule: \`src/utils/typography\`,
      },
    },
    {
      resolve: \`gatsby-transformer-remark\`,
      options: {
        plugins: [
          {
            resolve: \`gatsby-remark-images\`,
            options: {
              maxWidth: 590,
            },
          },
          {
            resolve: \`gatsby-remark-responsive-iframe\`,
            options: {
              wrapperStyle: \`margin-bottom: 1.0725rem\`,
            },
          },
          \`gatsby-remark-prismjs\`,
          \`gatsby-remark-copy-linked-files\`,
          \`gatsby-remark-smartypants\`,
        ],
      },
    },
    \`gatsby-plugin-sharp\`,
    {
      resolve: \`gatsby-plugin-google-analytics\`,
      options: {
        trackingId: \`UA-774017-3\`,
      },
    },
    {
      resolve: \`gatsby-plugin-manifest\`,
      options: {
        name: \`Bricolage\`,
        short_name: \`Bricolage\`,
        icon: \`static/logo.png\`,
        start_url: \`/\`,
        background_color: redish,
        theme_color: redish,
        display: \`minimal-ui\`,
      },
    },
    \`gatsby-plugin-offline\`,
    \`gatsby-plugin-react-helmet\`,
  ],
}
",
  "describe": "Install gatsby-plugin-foo in gatsby-config.js",
  "diff": "- Original  - 0
+ Modified  + 1

@@ -64,6 +64,7 @@
      },
      \`gatsby-plugin-offline\`,
      \`gatsby-plugin-react-helmet\`,
+     \\"gatsby-plugin-foo\\",
    ],
  }
",
  "id": "gatsby-plugin-foo",
  "name": "gatsby-plugin-foo",
  "newState": "const redish = \`#c5484d\`
module.exports = {
  siteMetadata: {
    title: \`Bricolage\`,
    author: \`Kyle Mathews\`,
    homeCity: \`San Francisco\`,
  },
  plugins: [
    {
      resolve: \`gatsby-source-filesystem\`,
      options: {
        path: \`\${__dirname}/src/pages\`,
        name: \`pages\`,
      },
    },
    \`gatsby-transformer-sharp\`,
    \`gatsby-plugin-emotion\`,
    {
      resolve: \`gatsby-plugin-typography\`,
      options: {
        pathToConfigModule: \`src/utils/typography\`,
      },
    },
    {
      resolve: \`gatsby-transformer-remark\`,
      options: {
        plugins: [
          {
            resolve: \`gatsby-remark-images\`,
            options: {
              maxWidth: 590,
            },
          },
          {
            resolve: \`gatsby-remark-responsive-iframe\`,
            options: {
              wrapperStyle: \`margin-bottom: 1.0725rem\`,
            },
          },
          \`gatsby-remark-prismjs\`,
          \`gatsby-remark-copy-linked-files\`,
          \`gatsby-remark-smartypants\`,
        ],
      },
    },
    \`gatsby-plugin-sharp\`,
    {
      resolve: \`gatsby-plugin-google-analytics\`,
      options: {
        trackingId: \`UA-774017-3\`,
      },
    },
    {
      resolve: \`gatsby-plugin-manifest\`,
      options: {
        name: \`Bricolage\`,
        short_name: \`Bricolage\`,
        icon: \`static/logo.png\`,
        start_url: \`/\`,
        background_color: redish,
        theme_color: redish,
        display: \`minimal-ui\`,
      },
    },
    \`gatsby-plugin-offline\`,
    \`gatsby-plugin-react-helmet\`,
    \\"gatsby-plugin-foo\\",
  ],
}
",
}
`;

exports[`gatsby-plugin resource e2e plugin resource test: GatsbyPlugin destroy 1`] = `undefined`;

exports[`gatsby-plugin resource e2e plugin resource test: GatsbyPlugin update 1`] = `
Object {
  "_message": "Installed gatsby-plugin-foo in gatsby-config.js",
  "id": "gatsby-plugin-foo",
  "name": "gatsby-plugin-foo",
  "options": undefined,
}
`;

exports[`gatsby-plugin resource e2e plugin resource test: GatsbyPlugin update plan 1`] = `
Object {
  "currentState": "const redish = \`#c5484d\`
module.exports = {
  siteMetadata: {
    title: \`Bricolage\`,
    author: \`Kyle Mathews\`,
    homeCity: \`San Francisco\`,
  },
  plugins: [
    {
      resolve: \`gatsby-source-filesystem\`,
      options: {
        path: \`\${__dirname}/src/pages\`,
        name: \`pages\`,
      },
    },
    \`gatsby-transformer-sharp\`,
    \`gatsby-plugin-emotion\`,
    {
      resolve: \`gatsby-plugin-typography\`,
      options: {
        pathToConfigModule: \`src/utils/typography\`,
      },
    },
    {
      resolve: \`gatsby-transformer-remark\`,
      options: {
        plugins: [
          {
            resolve: \`gatsby-remark-images\`,
            options: {
              maxWidth: 590,
            },
          },
          {
            resolve: \`gatsby-remark-responsive-iframe\`,
            options: {
              wrapperStyle: \`margin-bottom: 1.0725rem\`,
            },
          },
          \`gatsby-remark-prismjs\`,
          \`gatsby-remark-copy-linked-files\`,
          \`gatsby-remark-smartypants\`,
        ],
      },
    },
    \`gatsby-plugin-sharp\`,
    {
      resolve: \`gatsby-plugin-google-analytics\`,
      options: {
        trackingId: \`UA-774017-3\`,
      },
    },
    {
      resolve: \`gatsby-plugin-manifest\`,
      options: {
        name: \`Bricolage\`,
        short_name: \`Bricolage\`,
        icon: \`static/logo.png\`,
        start_url: \`/\`,
        background_color: redish,
        theme_color: redish,
        display: \`minimal-ui\`,
      },
    },
    \`gatsby-plugin-offline\`,
    \`gatsby-plugin-react-helmet\`,
    \\"gatsby-plugin-foo\\",
  ],
}
",
  "describe": "Install gatsby-plugin-foo in gatsby-config.js",
  "diff": "Compared values have no visual difference.",
  "id": "gatsby-plugin-foo",
  "name": "gatsby-plugin-foo",
  "newState": "const redish = \`#c5484d\`
module.exports = {
  siteMetadata: {
    title: \`Bricolage\`,
    author: \`Kyle Mathews\`,
    homeCity: \`San Francisco\`,
  },
  plugins: [
    {
      resolve: \`gatsby-source-filesystem\`,
      options: {
        path: \`\${__dirname}/src/pages\`,
        name: \`pages\`,
      },
    },
    \`gatsby-transformer-sharp\`,
    \`gatsby-plugin-emotion\`,
    {
      resolve: \`gatsby-plugin-typography\`,
      options: {
        pathToConfigModule: \`src/utils/typography\`,
      },
    },
    {
      resolve: \`gatsby-transformer-remark\`,
      options: {
        plugins: [
          {
            resolve: \`gatsby-remark-images\`,
            options: {
              maxWidth: 590,
            },
          },
          {
            resolve: \`gatsby-remark-responsive-iframe\`,
            options: {
              wrapperStyle: \`margin-bottom: 1.0725rem\`,
            },
          },
          \`gatsby-remark-prismjs\`,
          \`gatsby-remark-copy-linked-files\`,
          \`gatsby-remark-smartypants\`,
        ],
      },
    },
    \`gatsby-plugin-sharp\`,
    {
      resolve: \`gatsby-plugin-google-analytics\`,
      options: {
        trackingId: \`UA-774017-3\`,
      },
    },
    {
      resolve: \`gatsby-plugin-manifest\`,
      options: {
        name: \`Bricolage\`,
        short_name: \`Bricolage\`,
        icon: \`static/logo.png\`,
        start_url: \`/\`,
        background_color: redish,
        theme_color: redish,
        display: \`minimal-ui\`,
      },
    },
    \`gatsby-plugin-offline\`,
    \`gatsby-plugin-react-helmet\`,
    \\"gatsby-plugin-foo\\",
  ],
}
",
}
`;

exports[`gatsby-plugin resource handles config options as an object 1`] = `
Array [
  Object {
    "name": "gatsby-source-filesystem",
    "options": Object {
      "name": "pages",
      "path": "\${__dirname}/src/pages",
    },
  },
  Object {
    "name": "gatsby-transformer-sharp",
    "options": undefined,
  },
  Object {
    "name": "gatsby-plugin-emotion",
    "options": undefined,
  },
  Object {
    "name": "gatsby-plugin-typography",
    "options": Object {
      "pathToConfigModule": "src/utils/typography",
    },
  },
  Object {
    "name": "gatsby-transformer-remark",
    "options": Object {
      "plugins": Array [
        Object {
          "options": Object {
            "maxWidth": 590,
          },
          "resolve": "gatsby-remark-images",
        },
        Object {
          "options": Object {
            "wrapperStyle": "margin-bottom: 1.0725rem",
          },
          "resolve": "gatsby-remark-responsive-iframe",
        },
        "gatsby-remark-prismjs",
        "gatsby-remark-copy-linked-files",
        "gatsby-remark-smartypants",
      ],
    },
  },
  Object {
    "name": "gatsby-plugin-sharp",
    "options": undefined,
  },
  Object {
    "name": "gatsby-plugin-google-analytics",
    "options": Object {
      "trackingId": "UA-774017-3",
    },
  },
  Object {
    "name": "gatsby-plugin-manifest",
    "options": Object {
      "background_color": null,
      "display": "minimal-ui",
      "icon": "static/logo.png",
      "name": "Bricolage",
      "short_name": "Bricolage",
      "start_url": "/",
      "theme_color": null,
    },
  },
  Object {
    "name": "gatsby-plugin-offline",
    "options": undefined,
  },
  Object {
    "name": "gatsby-plugin-react-helmet",
    "options": undefined,
  },
  Object {
    "name": "gatsby-plugin-foo",
    "options": Object {
      "bar": "baz",
      "baz": "qux",
      "foo": 1,
      "otherStuff": Array [
        Object {
          "bar": Array [
            Object {
              "foo": "bar",
            },
          ],
          "foo": "bar2",
        },
      ],
    },
  },
]
`;
